dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/merge/RenameReferenceVisitor.java

```diff
@@ -0,0 +1,230 @@
+import org.eclipse.microprofile.openapi.models.parameters.Parameter;
+import org.eclipse.microprofile.openapi.models.parameters.RequestBody;
+import org.eclipse.microprofile.openapi.models.responses.APIResponse;
+import org.eclipse.microprofile.openapi.models.security.SecurityRequirement;
+import org.eclipse.microprofile.openapi.models.security.SecurityScheme;
+
+import io.openliberty.microprofile.openapi20.merge.MergeProcessor.NameType;
+import io.openliberty.microprofile.openapi20.merge.MergeProcessor.RenameHolder;
+import io.openliberty.microprofile.openapi20.utils.OpenAPIModelVisitor;
+import io.openliberty.microprofile.openapi20.utils.OpenAPIModelWalker.Context;
+
+/**
+ * Visitor which updates references within an OpenAPI model after things have been renamed
+ * <p>
+ * Updates:
+ * <ul>
+ * <li>references to objects in {@link OpenAPI#getComponents()}
+ * <li>references from links to operationIds
+ * <li>references to tags from operations
+ * </ul>
+ */
+public class RenameReferenceVisitor implements OpenAPIModelVisitor {
+
+    private final RenameHolder renameHolder;
+
+    /**
+     * Creates a RenameReferenceVisitor
+     * 
+     * @param renameHolder the {@code RenameHolder} holding the renames for the model
+     */
+    public RenameReferenceVisitor(RenameHolder renameHolder) {
+        this.renameHolder = renameHolder;
+    }
+
+    private void updateReference(Reference<?> referrer, NameType nameType, String prefix) {
+        String reference = referrer.getRef();
+        if (reference != null && reference.startsWith(prefix)) {
+            String oldName = reference.split("/")[3];
+            String newName = renameHolder.lookupName(nameType, oldName);
+            if (!oldName.equals(newName)) {
+                referrer.setRef(prefix + newName);
+            }
+        }
+    }
+
+    @Override
+    public Example visitExample(Context context, Example example) {
+        updateReference(example, NameType.EXAMPLES, "#/components/examples/");
+        return example;
+    }
+
+    @Override
+    public Example visitExample(Context context, String key, Example example) {
+        updateReference(example, NameType.EXAMPLES, "#/components/examples/");
+        return example;
+    }
+
+    @Override
+    public Header visitHeader(Context context, String key, Header header) {
+        updateReference(header, NameType.HEADERS, "#/components/headers/");
+        return header;
+    }
+
+    @Override
+    public Link visitLink(Context context, String key, Link link) {
+        if (link.getRef() != null) {
+            updateReference(link, NameType.LINKS, "#/components/links/");
+        } else {
+            link.setOperationId(renameHolder.lookupName(NameType.OPERATION_ID, link.getOperationId()));
+            link.setOperationRef(updateOperationRef(link.getOperationRef()));
+        }
+        return link;
+    }
+
+    private static final Pattern PATH_REF_PATTERN = Pattern.compile("#/paths/([^/]+)/.*");
```

Comment:
If the component strings above are constants then maybe this path pattern goes with them?

Reply:
Didn't move this as it's very tied to the method where it's used.